File: 二次开发库和测试源码/VB测试源码/vbexample/Form1.vb
Public Class Form1
    Private ret As Integer
    Private canisopen As Byte

    Private Sub Label1_Click(sender As Object, e As EventArgs) Handles Label1.Click

    End Sub

    Private Sub ComboBox4_SelectedIndexChanged(sender As Object, e As EventArgs) Handles ComboBox4.SelectedIndexChanged

    End Sub

    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click
        If canisopen = 0 Then
            ret = ucandll.ucanbusdll.CAN_OpenDevice(ComboBox1.SelectedIndex)
            If ret = 0 Then
                canisopen = 1
                Button1.Text = "关闭"

                Dim tmp As Integer
                Dim tm1, tm2 As Double
                tm1 = Val(ComboBox2.Text)
                tmp = tm1 * 1000
                Debug.Print(tm1)
                Debug.Print(ComboBox2.Text)
                Dim cancfg As ucandll.Can_Config
                cancfg = New ucandll.Can_Config()
                cancfg.model = ComboBox3.SelectedIndex
                Debug.Print(cancfg.model)
                cancfg.configs = 0        '初值
                cancfg.configs += &H0001  '接通内部匹配电阻
                cancfg.configs += &H0002  '开启离线唤醒模式
                cancfg.configs += &H0004  '开启自动重传
                cancfg.baudrate = tmp
                Debug.Print(tmp)
                ucandll.ucanbusdll.CAN_Init(ComboBox1.SelectedIndex, cancfg)
                tm2 = cancfg.baudrate
                Debug.Print(tm2)
            End If
        Else
            ucandll.ucanbusdll.CAN_CloseDevice(ComboBox1.SelectedIndex)
            canisopen = 0
            Button1.Text = "打开"
        End If


    End Sub

    Private Sub Button3_Click(sender As Object, e As EventArgs) Handles Button3.Click
        If canisopen = 0 Then
            Return
        End If

        Dim cmsg As ucandll.Can_Msg
        cmsg = New ucandll.Can_Msg()
        cmsg.data = New Byte(7) {}
        cmsg.ID = Convert.ToUInt32(TextBox1.Text, 16)
        If ComboBox4.SelectedIndex = 1 Then
            cmsg.ExternFlag = 1
        Else cmsg.ExternFlag = 0
        End If

        If ComboBox5.SelectedIndex = 1 Then
            cmsg.RemoteFlag = 1
        Else cmsg.RemoteFlag = 0
        End If
        cmsg.DataLen = Val(dtlenth.Text)
        If cmsg.DataLen > 8 Then cmsg.DataLen = 8
        Dim candt As TextBox
        candt = dt1
        For i As Integer = 0 To cmsg.DataLen - 1
            cmsg.data(i) = Convert.ToByte(candt.Text, 16)
            If i < 7 Then
                candt = DirectCast(Me.GetNextControl(candt, True), TextBox)
            End If
        Next
        ucandll.ucanbusdll.CAN_Transmit(ComboBox1.SelectedIndex, cmsg, 1, 100)

    End Sub

    Private Sub Timer1_Tick(sender As Object, e As EventArgs) Handles Timer1.Tick

        If canisopen = 0 Then
            Return
        End If

        Dim cmsg As New ucandll.Can_Msg()
        Dim mLen As Int32

        mLen = ucandll.ucanbusdll.CAN_Receive(ComboBox1.SelectedIndex, cmsg, 1, 100)
        If mLen <= 0 Then
            Return
        End If

        Dim lines(5) As String
        Dim strs As String
        lines(0) = Convert.ToString(cmsg.ID, 16)
        lines(1) = Convert.ToString(cmsg.TimeStamp, 10)
        If cmsg.ExternFlag = 1 Then
            lines(2) = "扩展帧"
        Else
            lines(2) = "标准帧"
        End If

        If cmsg.RemoteFlag = 1 Then
            lines(3) = "远程帧"
        Else
            lines(3) = "数据帧"
        End If
        lines(4) = Convert.ToString(cmsg.DataLen, 10)

        For i As Integer = 0 To cmsg.DataLen - 1
            Dim lens As Byte
            strs = Hex(cmsg.data(i))
            lens = Len(strs)
            If Len(strs) < 2 Then strs = "0" & strs
            lines(5) += strs + " "
        Next

        DataGridView1.Rows.Add(lines)

    End Sub


    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load

        ComboBox2.SelectedIndex = 0
        ComboBox3.SelectedIndex = 0
        ComboBox4.SelectedIndex = 0
        ComboBox5.SelectedIndex = 0
        Dim drives As Integer
        drives = ucandll.ucanbusdll.CAN_ScanDevice()
        If drives > 0 Then
            Dim i As Integer
            For i = 0 To drives - 1
                ComboBox1.Items.Add(i)
            Next
            ComboBox1.SelectedIndex = 0
        End If
        Timer1.Enabled = True
    End Sub

End Class

File: 二次开发库和测试源码/c#/canbusexp/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using ucanbus;
using System.Runtime.InteropServices;

//using System.Threading;

namespace canbusexp
{
    public partial class Form1 : Form
    {
        UInt32 devnmb;
        int canopened;
      //  private Timer rectimer;
      //  private AutoResetEvent RecEvent;
       // private TimerCallback RecTimerDelegate;
        public Form1()
        {
            int i,itm;
            string str;
            InitializeComponent();
            itm= ucanbusdll.CAN_ScanDevice();
            Console.Write("CAN_ScanDevice = {0}", itm);
            if (itm > 0)
            {
                for (i = 0; i < itm; i++)
                {
                    str=string.Format("通道{0}",i);
                    this.comboBox1.Items.Add(str);
                }
                this.comboBox1.SelectedIndex = 0;
            }
            this.comboBox2.SelectedIndex = 0;
            this.comboBox3.SelectedIndex = 0;
            this.comboBox4.SelectedIndex = 0;
            this.comboBox5.SelectedIndex = 0;
        }

        byte findhex_formstr(ref byte idx, byte len, String str)
        {
            char dain;
            while (idx < len)
            {
                dain = str[idx];
                if ((dain >= '0') && (dain <= '9')) return ((byte)(dain - '0'));
                else if ((dain >= 'a') && (dain <= 'f')) return ((byte)(dain + 10 - 'a'));
                else if ((dain >= 'A') && (dain <= 'F')) return ((byte)(dain + 10 - 'F'));
                else (idx)++;
            }
            return 0xff;
        }

        byte str2u8(byte[] data, String str)
        {
            byte ret;
            int da, db;
            byte i = 0,j = 0;
            byte[] buf=new byte[8];
            ret = (byte)str.Length;

            while (i < ret)
            {
                da = findhex_formstr(ref i, ret, str);
                if (da == 0xff) da = 0;
                i++;
                db = findhex_formstr(ref i, ret, str);
                if (db == 0xff) { buf[j] = (byte)(da); j++; }
                else { buf[j] = (byte)((da << 4) | db); j++; }
                i++;
                if (j >= 8) break;
            }
            buf.CopyTo(data,0);
            return j;
        }
        private void button1_Click(object sender, EventArgs e)
        {
            devnmb = (UInt32)comboBox1.SelectedIndex;
            Console.Write("SelectedIndex = {0}", devnmb);

            if (ucanbusdll.CAN_OpenDevice((UInt32)devnmb) != 0)
            {
                MessageBox.Show("Open device fault!", "Error!", MessageBoxButtons.OK, MessageBoxIcon.Error);
                return;
            }

            winitcan();
            canopened = 1;
            MessageBox.Show("Open device success!", "success!", MessageBoxButtons.OK, MessageBoxIcon.Information);
            timer1.Start();
        }
        private void winitcan()
        {
            int rtn;
            uint tmp;
            double tm2;
            Can_Config cancfg;
            cancfg = new Can_Config();
            cancfg.model = (uint)comboBox2.SelectedIndex;
            tmp = Convert.ToUInt32(comboBox5.Text, 10);
            cancfg.baudrate = tmp*1000;
            cancfg.configs = 0;
            cancfg.configs |= 0x0001;  //接通内部匹配电阻
            cancfg.configs |= 0x0002;  //开启离线唤醒模式
            cancfg.configs |= 0x0004;  //开启自动重传
            rtn = ucanbusdll.CAN_Init(devnmb,ref cancfg);
            tm2 = cancfg.baudrate / 1000.0;
            Console.Write("real baudrate is = {0}\r\n", tm2);
        }
        private void button2_Click(object sender, EventArgs e)
        {
            if (canopened == 0) return;
            ucanbusdll.CAN_CloseDevice(devnmb);
            canopened = 0;
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            canopened = 0;
            Console.Write("canopened = 0");
        }
        private void button3_Click(object sender, EventArgs e)
        {
            if (canopened == 0) return;
            uint i = 0;
            IntPtr pt = Marshal.AllocHGlobal(Marshal.SizeOf(typeof(Can_Msg)) * 100);
            Can_Msg[] cmsg = new Can_Msg[100];
            for (i = 0; i < 100; i++) { cmsg[i] = new Can_Msg(); }
            for (i = 0; i < 10; i++)
            {
                cmsg[i].data = new byte[8];
                cmsg[i].ID = Convert.ToUInt32(textBox1.Text, 16) + i;
                cmsg[i].DataLen = str2u8(cmsg[i].data, textBox2.Text);
                cmsg[i].ExternFlag = (byte)this.comboBox3.SelectedIndex;
                cmsg[i].RemoteFlag = (byte)this.comboBox4.SelectedIndex;
                IntPtr ptr = (IntPtr)((UInt32)pt + i * Marshal.SizeOf(typeof(Can_Msg)));
                Marshal.StructureToPtr(cmsg[i], ptr, false);
            }
            ucanbusdll.CAN_Transmit(devnmb, pt, 10, 100);
        }
        private void timer1_Tick(object sender, EventArgs e)
        {
            if (canopened == 0) return;
            int lenth, i, j;
            String[] str = new String[5];
            IntPtr pt = Marshal.AllocHGlobal(Marshal.SizeOf(typeof(Can_Msg)) * 100);
            Can_Msg[] rxmsg = new Can_Msg[100];
            lenth = ucanbusdll.CAN_Receive(devnmb,pt, 10, 100);
            for (i = 0; i < lenth; i++)
            {
                IntPtr ptr = (IntPtr)((UInt32)pt + i * Marshal.SizeOf(typeof(Can_Msg)));
                rxmsg[i] = (Can_Msg)Marshal.PtrToStructure(ptr, typeof(Can_Msg));
                Console.Write("receive ID:{0}  ", rxmsg[i].ID);

                str[0] = Convert.ToString(rxmsg[i].ID, 16);
                str[1] = Convert.ToString(rxmsg[i].TimeStamp, 10);
                if (rxmsg[i].ExternFlag == 1) str[2] = "扩展帧";
                else str[2] = "标准帧";
                if (rxmsg[i].RemoteFlag == 1) str[3] = "远程帧";
                else str[3] = "数据帧";
                str[4] = "";
                for (j = 0; j < rxmsg[i].DataLen; j++)
                {
                    str[4] += rxmsg[i].data[j].ToString("x2") + " ";
                }
                dataGridView1.Rows.Add(str);
            }
            Marshal.FreeHGlobal(pt);
        }
    }
}
